File: solutions/javascript/bfs/01Matrix.js
/* Single-sourced BFS solution, passes 49/50 test cases on LeetCode. A passing solution can be found later in this file. */

const directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
];

class CellDistancePair {
    constructor(row, col, distanceFromStart) {
        this.row = row;
        this.col = col;
        this.distanceFromStart = distanceFromStart;
    }

    getPositionString() {
        return `${this.row}, ${this.col}`;
    }
}

const create2dArrayOfSize = (numRows, numCols) => {
    const array = new Array(numRows);

    for (let row = 0; row < numRows; row++) {
        array[row] = new Array(numCols);
    }

    return array;
};

const getAdjacentPositions = (row, col) => {
    const adjacentPositions = [];

    for (const direction of directions) {
        const [rowChange, colChange] = direction;

        const adjacentPositionRow = row + rowChange;
        const adjacentPositionCol = col + colChange;

        adjacentPositions.push([adjacentPositionRow, adjacentPositionCol]);
    }

    return adjacentPositions;
};

const isInBounds = (matrix, row, col) => {
    const numRows = matrix.length;
    const numCols = matrix[0].length;

    const rowInBounds = row >= 0 && row < numRows;
    const colInBounds = col >= 0 && col < numCols;

    return rowInBounds && colInBounds;
};

const getNeighborPositions = (matrix, row, col) => {
    const adjacentPositions = getAdjacentPositions(row, col);
    const neighbors = [];

    for (const adjacentPosition of adjacentPositions) {
        const [adjacentPositionRow, adjacentPositionCol] = adjacentPosition;

        if (!isInBounds(matrix, adjacentPositionRow, adjacentPositionCol)) continue;

        neighbors.push(adjacentPosition);
    }

    return neighbors;
};

const getDistanceToNearestZero = (matrix, startRow, startCol) => {
    const queue = new Queue();
    const visited = new Set();

    const startCellDistancePair = new CellDistancePair(startRow, startCol, 0);
    queue.enqueue(startCellDistancePair);
    visited.add(startCellDistancePair.getPositionString());

    while (queue.size() > 0) {
        // Remove node
        const { row, col, distanceFromStart } = queue.dequeue();

        // Process node
        const curValue = matrix[row][col];
        if (curValue === 0) return distanceFromStart;

        // Add neighbors
        const neighborPositions = getNeighborPositions(matrix, row, col);
        for (const neighborPosition of neighborPositions) {
            const [neighborRow, neighborCol] = neighborPosition;

            const neighborCellDistancePair = new CellDistancePair(
                neighborRow,
                neighborCol,
                distanceFromStart + 1,
            );

            const neighborPositionString = neighborCellDistancePair.getPositionString();
            if (visited.has(neighborPositionString)) continue;

            visited.add(neighborPositionString);
            queue.enqueue(neighborCellDistancePair);
        }
    }
};

const updateMatrix = (matrix) => {
    const numRows = matrix.length;
    const numCols = matrix[0].length;

    const distanceToNearestZeroGrid = create2dArrayOfSize(numRows, numCols);

    for (let row = 0; row < numRows; row++) {
        for (let col = 0; col < numCols; col++) {
            const curValue = matrix[row][col];

            let distanceToNearestZero = 0;

            if (curValue === 1) {
                distanceToNearestZero = getDistanceToNearestZero(matrix, row, col);
            }

            distanceToNearestZeroGrid[row][col] = distanceToNearestZero;
        }
    }

    return distanceToNearestZeroGrid;
};

/* Multi-sourced BFS solution, passes on LeetCode */
const directions = [
    [-1, 0],
    [0, -1],
    [1, 0],
    [0, 1],
];

const getLocationString = (location) => {
    const { row, col } = location;

    return `${row}, ${col}`;
};

const getTargetCellLocations = (matrix, targetValue) => {
    const targetCellLocations = [];

    const numRows = matrix.length;
    const numCols = matrix[0].length;

    for (let row = 0; row < numRows; row += 1) {
        for (let col = 0; col < numCols; col += 1) {
            if (matrix[row][col] === targetValue) {
                targetCellLocations.push({
                    row,
                    col,
                });
            }
        }
    }

    return targetCellLocations;
};

const create2dArrayOfSize = (numRows, numCols, fillValue = undefined) => {
    const array = new Array(numRows);

    for (let row = 0; row < numRows; row += 1) {
        array[row] = new Array(numCols).fill(fillValue);
    }

    return array;
};

const isInBounds = (matrix, row, col) => {
    const numRows = matrix.length;
    const numCols = matrix[0].length;

    const isRowInBounds = row >= 0 && row < numRows;
    const isColInBounds = col >= 0 && col < numCols;

    return isRowInBounds && isColInBounds;
};

const getNeighborLocations = (matrix, row, col) => {
    const neighborLocations = [];

    for (const direction of directions) {
        const [rowChange, colChange] = direction;

        const newRow = row + rowChange;
        const newCol = col + colChange;

        if (!isInBounds(matrix, newRow, newCol) || matrix[newRow][newCol] === 0) continue;

        neighborLocations.push({ row: newRow, col: newCol });
    }

    return neighborLocations;
};

const updateMatrix = (matrix) => {
    if (matrix.length === 0) return matrix;

    const numRows = matrix.length;
    const numCols = matrix[0].length;

    const queue = new Queue();
    const visited = new Set();

    const distanceToNearestZeroMatrix = create2dArrayOfSize(numRows, numCols, 0);
    const zeroCellLocations = getTargetCellLocations(matrix, 0);

    for (const zeroCellLocation of zeroCellLocations) {
        queue.enqueue({
            ...zeroCellLocation,
            distanceFromZero: 0,
        });

        const zeroCellLocationString = getLocationString(zeroCellLocation);
        visited.add(zeroCellLocationString);
    }

    while (queue.size() > 0) {
        // Remove node
        const { row, col, distanceFromZero } = queue.dequeue();

        // Process node
        // We know that we have a 1 that is unvisited, which means we have just found the
        // length of the shortest path to that 1
        distanceToNearestZeroMatrix[row][col] = distanceFromZero;

        // Add neighbors
        const neighborLocations = getNeighborLocations(matrix, row, col);
        for (const neighborLocation of neighborLocations) {
            const neighborLocationString = getLocationString(neighborLocation);

            if (visited.has(neighborLocationString)) continue;
            visited.add(neighborLocationString);

            queue.enqueue({
                ...neighborLocation,
                distanceFromZero: distanceFromZero + 1,
            });
        }
    }

    return distanceToNearestZeroMatrix;
};

File: solutions/python/bfs/01matrix.py
"""
Single-sourced BFS solution, passes 49/50 test cases on LeetCode. A passing solution can be found later in this file.
"""
from collections import deque

directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
]

def create_2d_list_of_size(num_rows, num_cols, fill_value = None):
    return [[fill_value for _col in range(num_cols)] for _row in range(num_rows)]
    
def is_in_bounds(matrix, row, col):
    num_rows = len(matrix)
    num_cols = len(matrix[0])

    row_in_bounds = row >= 0 and row < num_rows
    col_in_bounds = col >= 0 and col < num_cols

    return row_in_bounds and col_in_bounds

class Solution:
    def get_adjacent_positions(self, row, col):
        adjacent_positions = []

        for direction in directions:
            row_change, col_change = direction

            adjacent_position_row = row + row_change
            adjacent_position_col = col + col_change

            adjacent_positions.append((adjacent_position_row, adjacent_position_col))

        return adjacent_positions

    def get_neighbor_positions(self, matrix, row, col):
        adjacent_positions = self.get_adjacent_positions(row, col)
        neighbors = []

        for adjacent_position in adjacent_positions:
            adjacent_position_row, adjacent_position_col = adjacent_position

            if not is_in_bounds(matrix, adjacent_position_row, adjacent_position_col):
                continue

            neighbors.append(adjacent_position)

        return neighbors

    def get_distance_to_nearest_zero(self, matrix, startRow, startCol):
        queue = deque()
        visited = set()

        start_position = (startRow, startCol, 0)
        
        queue.append(start_position)
        visited.add(start_position)

        while queue:
            # Remove node
            row, col, distance_from_start = queue.popleft()

            # Process node
            cur_value = matrix[row][col]
            if cur_value == 0:
                return distance_from_start

            # Add neighbors
            neighbor_position = self.get_neighbor_positions(matrix, row, col)
            for neighbor_position in neighbor_position:
                if neighbor_position in visited:
                    continue

                visited.add(neighbor_position)
                queue.append((*neighbor_position, distance_from_start + 1))


    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        num_rows = len(matrix)
        num_cols = len(matrix[0])

        distance_to_nearest_zero_grid = create_2d_list_of_size(num_rows, num_cols)

        for row in range(num_rows):
            for col in range(num_cols):
                cur_value = matrix[row][col]

                distance_to_nearest_zero = 0

                if cur_value == 1:
                    distance_to_nearest_zero = self.get_distance_to_nearest_zero(
                        matrix, row, col)

                distance_to_nearest_zero_grid[row][col] = distance_to_nearest_zero

        return distance_to_nearest_zero_grid

"""
Multi-sourced BFS solution, passes on LeetCode.
"""
from collections import deque

directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
]

def create_2d_list_of_size(num_rows, num_cols, fill_value = None):
    return [[fill_value for _col in range(num_cols)] for _row in range(num_rows)]

def is_in_bounds(matrix, row, col):
    num_rows = len(matrix)
    num_cols = len(matrix[0])

    row_in_bounds = row >= 0 and row < num_rows
    col_in_bounds = col >= 0 and col < num_cols

    return row_in_bounds and col_in_bounds

class Solution:
    def get_cells_with_zero(self, matrix, num_rows, num_cols):
        cells_with_zero = []

        for row in range(num_rows):
            for col in range(num_cols):
                cell = matrix[row][col]

                if cell == 0:
                    cells_with_zero.append([row, col])

        return cells_with_zero

    def get_neighbors(self, matrix, row, col):
        neighbors = []

        for direction in directions:
            row_change, col_change = direction

            new_row = row + row_change
            new_col = col + col_change

            if not is_in_bounds(matrix, new_row, new_col): continue

            neighbors.append((new_row, new_col))

        return neighbors


    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        num_rows = len(matrix)
        num_cols = len(matrix[0])

        distance_to_nearest_zero = create_2d_list_of_size(num_rows, num_cols, float("inf"))

        cells_with_zero = self.get_cells_with_zero(matrix, num_rows, num_cols)

        queue = deque()
        visited = set()

        for cell in cells_with_zero:
            row, col = cell

            cell_tuple = (row, col, 0)
            queue.append(cell_tuple)

            visited.add(cell_tuple)

        while queue:
            # Remove node
            row, col, distance_so_far = queue.popleft()

            # Process node
            if matrix[row][col] == 0:
                distance_to_nearest_zero[row][col] = 0
            elif matrix[row][col] == 1:
                distance_to_nearest_zero[row][col] = distance_so_far

            # Add neighbors
            neighbors = self.get_neighbors(matrix, row, col)
            for neighbor in neighbors:
                if neighbor in visited:
                    continue
                visited.add(neighbor)

                queue.append((*neighbor, distance_so_far + 1))

        return distance_to_nearest_zero
